### PR TITLE
#7739 

### PR DESCRIPTION
**Pull request by HeorhiiBoliuk**
Created a custom registration form that include a extra fields for registration of user. Created a service for saving extra fields in custom db table. Created hook_entity_insert for getting a values of created user.

A db dump was added cause the last dump was updated a long time ago.

### CODE REVIEW

**Review by ivandudardb — APPROVED** [review]
LGTM

File: web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  /**
+   * {@inheritDoc}
+   */
+  public function form(array $form, FormStateInterface $form_state): array {
+
+    $form = parent::form($form, $form_state);
+    $cities = $this->cityService->getCitiesArray();
+    $term_options = $this->getTaxonomyTerm();
+
+    $form['password'] = [
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php:88]: isn't password declared in parent form?

File: web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state): void {
+    $username = $form_state->getValue('name');
+    $email = $form_state->getValue('mail');
+    $password = $form_state->getValue('password');
+    $this->entity->setUsername($username);
+    $this->entity->setEmail($email);
+    $this->entity->setPassword($password);
+    $this->entity->activate();
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php:152]: This should've been handled by the parent form.

File: web/modules/custom/registration_form_custom/registration_form_custom.module [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\registration_form_custom\Form\NewUserRegisterForm;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\weather_block\Services\UserWeatherHandler;
+
+/**
+ * Implements hook_entity_insert().
+ */
+function registration_form_custom_entity_insert(\Drupal\user\Entity\User $entity) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.module:11]: Why do you need a hook when you have a custom form with all of that there in the same module?

File: web/modules/custom/registration_form_custom/registration_form_custom.module [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  $user_id = $entity->id();
+  $city_value = \Drupal::request()->request->get('city');
+  $country_value = \Drupal::request()->request->get('country');
+  $interested_value = \Drupal::request()->request->get('interested');
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.module:15]: While form API will clear requests from values if validation fails, it is still not secure to use them in this way.

File: web/modules/custom/registration_form_custom/registration_form_custom.info.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+name: Registration Form Custom
+type: module
+description: 'Custom registration form'
+core_version_requirement: ^9 || ^10
+package: Custom
+dependencies:
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.info.yml:6]: I saw the use of weather module in this module, but it isn't in the dependencies

File: web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    $country_options = [];
+    $countries = $this->countryManager->getList();
+    foreach ($countries as $code => $name) {
+      $country_options[$code] = $name;
+    }
+    return $country_options;
+  }
+
+  /**
+   * Get a taxonomy terms.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php:129]: Could've been more descriptive.
You are clearly getting only news terms

File: web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      '#title' => $this->t('Choose your city'),
+      '#options' => array_combine($cities, $cities),
+      '#empty_option' => $this->t('-select-'),
+      '#description' => $this->t('Select your city'),
+    ];
+    $form['interested'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Interests'),
+      '#options' => $term_options,
+      '#description' => $this->t('Select your interests'),
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php:111]: I guess this allows multi selection, like description suggests, right?

File: web/modules/custom/registration_form_custom/registration_form_custom.install [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+use Drupal\Core\Database\Database;
+use Drupal\Core\Database\Schema;
+
+/**
+ * Implements hook_schema().
+ */
+function registration_form_custom_schema() {
+  $schema['extra_field_register'] = [
+    'description' => 'Stores the default city for users who did not select any city.',
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.install:15]: Does it?

File: web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace Drupal\registration_form_custom\Services;
+
+use Drupal\Core\Database\Connection;
+
+/**
+ * Class methods adds extra fields in custom DB.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php:8]: are they?

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      'attributes' => [
+        'class' => [
+          'use-ajax',
+          'register-popup-form',
+        ],
+        'data-dialog-type' => 'modal',
+      ],
+    ];
+    $url->setOptions($link_options);
+    $link = Link::fromTextAndUrl(t('Register'), $url)->toString();
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:35]: DI

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    $build = [];
+    if (\Drupal::currentUser()->isAnonymous()) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:37]: DII

File: web/modules/custom/registration_form_custom/registration_form_custom.install [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    'fields' => [
+      'uid' => [
+        'type' => 'serial',
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.install:18]: Should it be serial?

File: web/modules/custom/registration_form_custom/registration_form_custom.install [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        'unsigned' => TRUE,
+        'not null' => TRUE,
+        'description' => 'Primary Key: Unique ID for the user.',
+      ],
+      'country' => [
+        'type' => 'varchar',
+        'length' => 255,
+        'not null' => TRUE,
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/registration_form_custom.install:26]: not null
and not required in the form
Something should be changed.

File: web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class ExtraFieldsHandler {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php:10]: FYI extra fields are functionality that exists in Drupal, so the name is misleading (it doesn't describe what this service does as well).
But you can leave it.

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

File: web/modules/custom/registration_form_custom/registration_form_custom.install [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        'description' => 'Name of the selected country.',
+      ],
+      'interested' => [
+        'type' => 'varchar',
+        'length' => 255,
+        'not null' => TRUE,
+        'description' => 'Name of the selected category name.',
+      ],
+    ],
+    'primary key' => ['uid'],
```
[the remark this answers is not in the document]
Comment (HeorhiiBoliuk) [reply web/modules/custom/registration_form_custom/registration_form_custom.install:36]: Why? #Non-default fields should be stored in the custom database table. 
Non-default fields is country, city, interested in the rest of fields is saving in default DB or I'm don't get it?

File: web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php (hunk 1 above)
Comment (lavrenchuk-dev) [reply web/modules/custom/registration_form_custom/src/Form/NewUserRegisterForm.php:87]: image
/admin/config/people/accounts

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+ *   admin_label = @Translation("Register form popup"),
+ * )
+ */
+class RegisterFormPopup extends BlockBase implements ContainerFactoryPluginInterface {
+  public function __construct(array $configuration,
+  $plugin_id,
+  $plugin_definition,
+    protected LinkGeneratorInterface $link_generator,
+  protected AccountProxyInterface $current_user,
+  protected UrlGeneratorInterface $url_generator) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:27]: formatting died

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    $url->setOptions([
+      'attributes' => [
+        'class' => ['use-ajax', 'register-popup-form'],
+        'data-dialog-type' => 'modal',
+      ],
+    ]);
+    $link = $this->link_generator->generate(t('Register'), $url);
+    $build = [];
+
+    if ($this->current_user->isAnonymous()) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:59]: Does this work properly with cache enabled?

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  public function build() {
+    $url = Url::fromRoute('user.register');
+    $url->setOptions([
+      'attributes' => [
+        'class' => ['use-ajax', 'register-popup-form'],
+        'data-dialog-type' => 'modal',
+      ],
+    ]);
+    $link = $this->link_generator->generate(t('Register'), $url);
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:56]: DI

File: web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * Initialize a service property.
+   */
+  public function __construct(protected Connection $database) {
+  }
+
+  /**
+   * This function is saving extra fields that i`m added to default register
+   * form.
```
[the remark this answers is not in the document]
Comment (lavrenchuk-dev) [reply web/modules/custom/registration_form_custom/src/Services/ExtraFieldsHandler.php:21]: Still wrong standarts

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+use Drupal\Core\Routing\UrlGeneratorInterface;
+use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\Url;
+use Drupal\Core\Utility\LinkGeneratorInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a block for the popup registration form.
+ *
+ * @Block(
```
Comment (lavrenchuk-dev) [inline web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:16]: Use attributes

File: web/modules/custom/registration_form_custom/registration_form_custom.install (hunk 15 above)
[the remark this answers is not in the document]
Comment (lavrenchuk-dev) [reply web/modules/custom/registration_form_custom/registration_form_custom.install:36]: There is no city in schema
I understand you have another table
I meant to move it here

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]

File: web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php (hunk 17 above)
Comment (HeorhiiBoliuk) [reply web/modules/custom/registration_form_custom/src/Plugin/Block/RegisterFormPopup.php:59]: Yes, enabled a caches and block is showing for unregister user

**Review by lavrenchuk-dev — APPROVED** [review]

**Pull request merged**